### PR TITLE
Add elasticsearch alerting engine

### PR DESCRIPTION
**Pull request by beardface**
#5893 

This branch introduces a, hopefully, lighter-weight variant of an elasticsearch alerting engine for Grafana.

The other approaches I've seen seem very heavy on implementation; and internal ES logic.  This approach attempts to take what I consider to be a simpler/cleaner path.

It is working well for us; so I figured I'd share in the event its useful enough to merge; or others would like to leverage the implementation.

### CODE REVIEW

**Comment by torkelo** [comment]
looks like a good start, do we really need the elastic go client lib? 

Also the response handling looks a bit simplistic, does it support multiple levels of bucket aggregations? For example if you have a terms aggregation followed by a date histogram aggregation.

**Comment by m2b** [comment]
Without looking at other requests, I am hoping this is not too complicated.  You already compute metrics quite nicely from Elastic.  All we need is the capability to define alerts on the panels to work.  Checking against limits, etc.

**Comment by beardface** [comment]
@torkelo I had pulled in the es client lib b/c it simplified unmarshalling; but I really dislike having it in here... I'm going to pull it out this morning. Since its pretty bloated for what we need here.

**Comment by beardface** [comment]
@torkelo Can I get a unit test case that covers this `Also the response handling looks a bit simplistic, does it support multiple levels of bucket aggregations? For example if you have a terms aggregation followed by a date histogram aggregation.` ?

If so, I'm happy to extend it.

Also; I put in a version check for elastic `2.x`... not sure if we need that though; this might support ex `5.x`, but I don't have an active instance to test on... thoughts there?

File: pkg/tsdb/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	result := &tsdb.BatchResult{}
+	result.QueryResults = make(map[string]*tsdb.QueryResult)
+
+	for _, q := range context.Queries {
+		esVersion, err := q.DataSource.JsonData.Get("esVersion").Int()
+		if err != nil {
+			return result.WithError(err)
+		}
+
+		if esVersion != 2 {
```
Comment (beardface) [inline pkg/tsdb/elasticsearch/elasticsearch.go]: Thoughts here?

**Comment by torkelo** [comment]
Here is a test case with one terms and one date histogram aggregation, (you can have multiple terms & filter aggregations), each one will nest the result even more, so the response handling needs to handle the recursion. 

````
targets = [{
        refId: 'A',
        metrics: [{type: 'count', id: '1'}, {type: 'avg', field: '@value', id: '4'}],
        bucketAggs: [{type: 'terms', field: 'host', id: '2'}, {type: 'date_histogram', field: '@timestamp', id: '3'}],
      }];
      response =  {
        responses: [{
          aggregations: {
            "2": {
              buckets: [
              {
                "3": {
                  buckets: [
                  { "4": {value: 10}, doc_count: 1, key: 1000},
                  { "4": {value: 12}, doc_count: 3, key: 2000}
                  ]
                },
                doc_count: 4,
                key: 'server1',
              },
              {
                "3": {
                  buckets: [
                  { "4": {value: 20}, doc_count: 1, key: 1000},
                  { "4": {value: 32}, doc_count: 3, key: 2000}
                  ]
                },
                doc_count: 10,
                key: 'server2',
              },
              ]
            }
          }
        }]
      };
```

**Comment by beardface** [comment]
@torkelo I added in the test case you provided and extended this to support nested buckets via a recursive processing loop.

I also simplified the types even more and got rid of the need to import anything at all from elastic (or copy paste... which was arguably worse. :) )

**Comment by torkelo** [comment]
This this PR use the same series name logic (and alias patterns) as the frontend code?

**Comment by beardface** [comment]
@torkelo ah; no - missed that.  Adding now.

**Comment by beardface** [comment]
@torkelo Ok, I added in logic to compute the preferred names based on the logic from the frontend; did a few tests on my end and is looking good here.

Let me know what you think.

Thanks!

**Comment by torkelo** [comment]
this is looking really good, will try to merge soon.

File: pkg/tsdb/query_context.go [hunk 2]
```diff
@@ -6,9 +6,9 @@ type QueryContext struct {
 	TimeRange   *TimeRange
 	Queries     QuerySlice
 	Results     map[string]*QueryResult
-	ResultsChan chan *BatchResult
-	Lock        sync.RWMutex
-	BatchWaits  sync.WaitGroup
+	ResultsChan chan *BatchResult `json:"-"`
+	Lock        sync.RWMutex      `json:"-"`
+	BatchWaits  sync.WaitGroup    `json:"-"`
```
Comment (torkelo) [inline pkg/tsdb/query_context.go]: is there a reason for adding these, when is QueryContext ever marshalled to json ?

Comment (beardface) [inline pkg/tsdb/query_context.go]: Ah, I was initially marshaling the hole thing; but I stopped doing that a while ago; good catch.

**Comment by proffalken** [comment]
Following, would love to see this merged and working :)

**Comment by beardface** [comment]
@torkelo I'm doing some more testing today and it looks like it has problems when you use template variables; but I don't see the template variables getting passed into the context so that I can use it to do replacements...

Is that a known bug?

Is there some way to get the templated variables in `Execute` so that I can do query replacements when sending the request to elasticsearch?

**Comment by beardface** [comment]
Upon further thought; that's probably not possible to do easily - or could lead to more confusion when people change their template variable.

Just trying to see what the limitations are and if I missed anything.

**Comment by torkelo** [comment]
template variables does not need to be supported beside the two built in `$__intervalMS` and `__interval`  (see InfluxDB backend code for reference)

**Comment by beardface** [comment]
@torkelo based off the code; I presume you meant `$__interval_ms` instead of `$__intervalMS`.

I've updated the logic to handle that case.

**Comment by torkelo** [comment]
Yes, I meant that,  great that you fixed that, will merge early next week

**Comment by torkelo** [comment]
Got around to testing this PR today. 

1) Got error that Elasticsearch 5 is not supported

Switched 2 Elasticsearch 2.4 to test further and got this error: 

```
tsdb.HandleRequest() error Failed to get metrics: "{\"error\":{\"root_cause\":[{\"type\":\"index_not_found_exception\",\"reason\":\"no such index\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"metrics-2017.02.22\",\"index\":\"metrics-2017.02.22\"}],\"type\":\"index_not_found_exception\",\"reason\":\"no such index\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"metrics-2017.02.22\",\"index\":\"metrics-2017.02.22\"},\"status\":404}" (404 Not Found)
```

The frontend is building this query
```
{"search_type":"count","ignore_unavailable":true,"index":["metrics-2017.02.21"]}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1487670622085","lte":"1487671522085","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"@metric:logins.count"}}]}},"aggs":{"2":{"date_histogram":{"interval":"10s","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1487670622085","max":"1487671522085"},"format":"epoch_millis"},"aggs":{"1":{"avg":{"field":"@value"}}}}}}
```

Maybe the index patterns in the backend is not using UTC when generating the daily index name? Or some other reason why the backend day is ahead by 1 ?

**Comment by torkelo** [comment]
in order for this PR not to get too big and for others to test it more we decided to merge it into a branch, https://github.com/grafana/grafana/tree/alerting-elasticsearch  

please issue bug fixes and further work to that branch.

**Comment by m2b** [comment]
I am a bit new on open source.  Should I pull this branch down?  How do I
build it?  Are the binaries?
https://github.com/grafana/grafana/tree/alerting-elasticsearch

**Pull request closed without merging**